nightshift/datastore.py

```diff
@@ -0,0 +1,205 @@
+# -*- coding: utf-8 -*-
+
+"""
+NightShift's database schema.
+
+
+dialect+driver://username:password@host:port/database
+# psycopg2
+engine = create_engine('postgresql+psycopg2://scott:tiger@localhost/mydatabase', client_encoding='utf8')
+"""
+from datetime import datetime
+
+from sqlalchemy import (
+    Boolean,
+    Column,
+    Date,
+    DateTime,
+    ForeignKey,
+    Integer,
+    PickleType,
+    PrimaryKeyConstraint,
+    String,
+    UniqueConstraint,
+)
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy.orm import relationship
+
+
+Base = declarative_base()
+
+
+class Library(Base):
+    """
+    Library system.
+    """
+
+    __tablename__ = "library"
+
+    nid = Column(Integer, primary_key=True)
+    code = Column(String(3), unique=True)
+
+    def __repr__(self):
+        return f"<Library(nid='{self.nid}', code='{self.code}')>"
+
+
+class OutputFile(Base):
+    """
+    Output MARC file info.
+    """
+
+    __tablename__ = "output_file"
+    __table_args__ = (UniqueConstraint("handle", "libraryId"),)
+
+    nid = Column(Integer, primary_key=True)
+    libraryId = Column(Integer, ForeignKey("library.nid"), nullable=False)
+    handle = Column(String, nullable=False)
+    timestamp = Column(DateTime, default=datetime.now())
+
+    def __repr__(self):
+        return (
+            f"<OutputFile(nid='{self.nid}', "
+            f"libraryId='{self.libraryId}', "
+            f"handle='{self.handle}', "
+            f"timestamp='{self.timestamp}')>"
+        )
+
+
+class Resource(Base):
+    """
+    Resource to be upgraded info.
+    """
+
+    __tablename__ = "resource"
+    __table_args__ = (
+        PrimaryKeyConstraint("sierraId", "libraryId"),
+        {},
+    )
+
+    sierraId = Column(Integer, nullable=False)
+    libraryId = Column(Integer, ForeignKey("library.nid"), nullable=False)
+    resourceCategoryId = Column(
+        Integer, ForeignKey("resource_category.nid"), nullable=False
+    )
+
+    archived = Column(Boolean, default=False)
+    bibDate = Column(Date)
+    author = Column(String(100, collation="uft8"))
+    title = Column(String(100, collation="utf8"))
+    pubDate = Column(String)
+
+    congressNumber = Column(String)
+    controlNumber = Column(String)
+    distributorNumber = Column(String)
+    otherNumber = Column(String)
+    outputId = Column(Integer, ForeignKey("output_file.nid"))
+    sourceId = Column(Integer, ForeignKey("source_file.nid"), nullable=False)
+    srcFieldsToKeep = Column(PickleType)
```

Comment:
Would it be beneficial to have this as a JSON object? I don`t think `PickleType` is queryable at all, but I'm assuming there might be benefits to having the db return a native object.

Reply:
The problem with Metadata API is that for searches it returns JSON but for getting a full bib it returns XML.
NightShift will need to search first, find OCLC number of the best match and then get full bib. I thought this pickling `request.Response` objects will work with both formats.